lib/CloudinaryPlugin.js

```diff
@@ -0,0 +1,333 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2019 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const Cloudinary = require('cloudinary').v2;
+
+class CloudinaryPlugin {
+
+  constructor() {
+    this.context = null;
+  }
+
+  init(config, context) {
+    this.context = context;
+
+    this.controllers = {
+      assetsController: {
+        search: 'search',
+        rename: 'rename',
+        destroy: 'destroy',
+        transform: 'transform'
+      },
+
+      tagsController: {
+        add_tag: 'add_tag',
+        remove_tag: 'remove_tag',
+        remove_all_tags: 'remove_all_tags',
+        replace_tag: 'replace_tag'
+      },
+    };
+
+    this.routes = [
+      {
+        verb: 'get',
+        url: '/assets/:expression',
+        controller: 'assetsController',
+        action: 'search'
+      },
+      {
+        verb: 'post',
+        url: '/assets/transform',
```

Comment:
nitpicking: as far as I understand, the `public_id` parameter is part of the definition of the resource. I would include it as an url parameter.

Reply:
Well i thought it was a better idea not to include it because Cloudinary allows special characters (espacially slashes and whitespace) in the public_id. These characters must be encoded if passed as an url parameter. I thought it was more convenient for the user to use the request body instead. (eg. if you want to transform your `my new sheep` asset which is in the `animal` directory, you must use `animal/my new sheep` as public_id and i found it confusing to put this as an url parameter)

Reply:
Ok, this is not very RESTful but the slash issue alone is enough to keep it as-is.